File: pkg/geo/bounds.go
package geo

import "math"

// Bounds is an axis-aligned bounding box ("AABB"). It is defined by its minimum
// and maximum points.
type Bounds [2]Vec

// NewBounds constructs a new AABB from the two points given.
func NewBounds(p1, p2 Vec) *Bounds {
	return &Bounds{vecMin(p1, p2), vecMax(p1, p2)}
}

// Intersect tests if the ray intersects the bounds. If it does, it returns the
// two t values in ascending order and the value true. Otherwise it returns
// false and garbage t values. Always check the returned boolean.
//
// https://www.scratchapixel.com/lessons/3d-basic-rendering/minimal-ray-tracer-rendering-simple-shapes/ray-box-intersection
func (b *Bounds) Intersect(ray *Ray) (t0, t1 float64, found bool) {
	return
}

// return the vector that is the component-wise minimum of the two vectors
func vecMin(a, b Vec) Vec {
	return Vec{
		math.Min(a.X, b.X),
		math.Min(a.Y, b.Y),
		math.Min(a.Z, b.Z),
	}
}

// return the vector that is the component-wise maximum of the two vectors
func vecMax(a, b Vec) Vec {
	return Vec{
		math.Max(a.X, b.X),
		math.Max(a.Y, b.Y),
		math.Max(a.Z, b.Z),
	}
}

File: src/geo/bounds.rs
use super::{Component, Point, Ray};
use crate::Float;
use std::mem;

/// An axis-aligned bounding box.
#[derive(Debug)]
pub struct Bounds {
    min: Point,
    max: Point,
}

impl Bounds {
    /// Create a new bounds from the given corner points.
    pub fn from_corners(p1: Point, p2: Point) -> Self {
        Self {
            min: Point::min(p1, p2),
            max: Point::max(p1, p2),
        }
    }

    /// Test a ray for intersection.
    ///
    /// If intersection is found, returns the `(t_near, t_far)` ray parameter
    /// values.
    pub fn intsersects(&self, ray: &Ray, t_min: Float, t_max: Float) -> Option<(f64, f64)> {
        // https://raytracing.github.io/books/RayTracingTheNextWeek.html#boundingvolumehierarchies/rayintersectionwithanaabb
        let (t0, t1) = Component::XYZ.iter().fold((t_min, t_max), |(t0, t1), &i| {
            let inv_ray_dir = ray.direction[i].recip();
            let mut t_near = (self.min[i] - ray.origin[i]) * inv_ray_dir;
            let mut t_far = (self.max[i] - ray.origin[i]) * inv_ray_dir;

            if t_near > t_far {
                mem::swap(&mut t_near, &mut t_far);
            }

            (t0.max(t_near), t1.min(t_far))
        });

        if t0 > t1 {
            None
        } else {
            Some((t0, t1))
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::geo::Vector;
    use super::*;

    #[test]
    fn intersects() {
        let bounds = Bounds::from_corners(Point::splat(-1.0), Point::splat(1.0));

        let ray = Ray::new(Point::new(0.0, 0.0, -10.0), Vector::Z_AXIS);
        assert_eq!(Some((9.0, 11.0)), bounds.intsersects(&ray, 0.0, Float::INFINITY));

        let ray = Ray::new(Point::new(0.0, 0.0, -10.0), Vector::Y_AXIS);
        assert_eq!(None, bounds.intsersects(&ray, 0.0, Float::INFINITY));
    }
}